### PR TITLE
Phase models are now part of flash and not properties.

### PR DESCRIPTION
**Pull request by TotoGaz**
Better separation of data and algorithms.
Removed some artificially `static` functions; better encapsulation of flash `const` states.

### CODE REVIEW

**Comment by TotoGaz** [comment]
@francoishamon can you have a look? Just a small refactoring this time.

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp [hunk 1]
```diff
@@ -139,7 +133,7 @@ std::vector< std::vector< std::vector< double > > > TableReader::readTables( std
 
 TableReader::Properties TableReader::buildTables( const std::vector< pvt::PHASE_TYPE > & phases,
                                                   const std::vector< std::string > & tableFileNames,
-                                                  const std::vector< double > & surfaceDensities,
+                                                  const std::vector< double > & surfaceMassDensities,
```
Comment (francoishamon) [inline PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:136]: Good catch, that is clearer now

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp [hunk 2]
```diff
@@ -157,25 +151,25 @@ TableReader::Properties TableReader::buildTables( const std::vector< pvt::PHASE_
...
-        result.gasSurfaceMassDensity = surfaceDensity;
+        result.gasSurfaceMassDensity = surfaceMassDensity;
         result.gasSurfaceMolecularWeight = molarWeight;
         break;
       case pvt::PHASE_TYPE::LIQUID_WATER_RICH:
         ASSERT( phaseTable.size() == 1, "Too many lines in water properties table" );
         result.waterTable = phaseTable[0];
-        result.waterSurfaceMassDensity = surfaceDensity;
+        result.waterSurfaceMassDensity = surfaceMassDensity;
         result.waterSurfaceMolecularWeight = molarWeight;
```
Comment (francoishamon) [inline PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:173]: Another nitpicking comment about naming for a future PR: it noticed that `molarWeight` and `molecularWeight` are used interchangeably to designate the same thing. Maybe we could use `molarWeight` everywhere as in the input file?

**Review by francoishamon — APPROVED** [review]
Looks good to me. Thanks! I just added to small name change suggestions for future refactorings.

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp (hunk 2 above)
Comment (TotoGaz) [reply PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:173]: OK, I noticed it but did not know which choice was more appropriate.

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp [hunk 3]
```diff
@@ -26,91 +26,85 @@ bool MultiphaseSystem::hasSucceeded() const
...
   {
-    sysProps.setPhaseMoleFractionDP( phase, dPhaseMoleFraction );
-    sysProps.setMolecularWeightDP( phase, dMolecularWeight );
-    sysProps.setMoleDensityDP( phase, dMoleDensity );
-    sysProps.setMassDensityDP( phase, dMassDensity );
-    sysProps.setMoleCompositionDP( phase, dMoleComposition );
-  };
-
-  updateDerivativeFiniteDifference( sysProps, diffed, dPressure, callBack );
+    const FiniteDifferenceValues values = updateDerivativeFiniteDifference( phase, sysProps, diffed, dPressure );
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:79]: `sysProps` will change but I've added `perturbed` which is a good suggestion. I'll keep it.

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp (hunk 1 above)
Comment (TotoGaz) [reply PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:136]: Thx.

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp [hunk 4]
```diff
@@ -26,91 +26,85 @@ bool MultiphaseSystem::hasSucceeded() const
...
+ * @param phase The phase for which we compute the finite difference
  * @param reference Reference solution.
  * @param variation Variation solution.
  * @param delta Finite difference step.
- * @param store Callback that will be called with the computed derivatives.
+ * @return A structure holding the finite difference computed values.
  *
  * Generic functions (pure code factorisation) that evaluates the derivatives
  * through finite differences process between reference @p sysProps and variation @p diffed.
  * This function then calls @p consumer, which role should mainly to store the computed data where they should be.
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:48]: Yes, I did not see the bottom part 🤷

File: PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp [hunk 5]
```diff
@@ -26,91 +26,85 @@ bool MultiphaseSystem::hasSucceeded() const
   return m_stateIndicator == State::SUCCESS;
 }
 
+struct FiniteDifferenceValues
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply PVTPackage/source/MultiphaseSystem/MultiphaseSystem.cpp:29]: done

**Pull request merged**